micro-rdk/src/common/provisioning/server.rs

```diff
@@ -0,0 +1,528 @@
+            SetSmartMachineCredentialsResponse,
+        },
+    },
+};
+
+use super::storage::Storage;
+
+struct ProvisioningServiceBuilder {
+    last_connection_attempt: Option<NetworkInfo>,
+    provisioning_info: Option<ProvisioningInfo>,
+    reason: ProvisioningReason,
+}
+
+impl ProvisioningServiceBuilder {
+    fn new() -> Self {
+        Self {
+            reason: ProvisioningReason::Erased,
+            last_connection_attempt: Default::default(),
+            provisioning_info: Default::default(),
+        }
+    }
+    fn with_provisioning_info(mut self, info: ProvisioningInfo) -> Self {
+        let _ = self.provisioning_info.insert(info);
+        self
+    }
+    fn with_reason(mut self, reason: ProvisioningReason) -> Self {
+        self.reason = reason;
+        self
+    }
+    fn with_network_info(mut self, info: NetworkInfo) -> Self {
+        let _ = self.last_connection_attempt.insert(info);
+        self
+    }
+    fn build<S: Storage + Clone>(self, storage: S) -> ProvisioningService<S> {
+        ProvisioningService {
+            provisioning_info: Rc::new(self.provisioning_info),
+            last_connection_attempt: Rc::new(self.last_connection_attempt),
+            reason: Rc::new(self.reason),
+            storage,
+            credential_ready: AtomicSync::default(),
+        }
+    }
+}
+
+#[derive(PartialEq)]
+enum ProvisioningReason {
+    Erased,
+    InvalidCredentials,
+}
+
+#[derive(Default)]
+struct NetworkInfo(provisioning::v1::NetworkInfo);
+#[derive(Default)]
+struct ProvisioningInfo(provisioning::v1::ProvisioningInfo);
+
+impl ProvisioningInfo {
+    fn set_fragment_id(&mut self, frag_id: String) {
+        self.0.fragment_id = frag_id;
+    }
+    fn set_model(&mut self, model: String) {
+        self.0.model = model;
+    }
+    fn set_manufacturer(&mut self, manufacturer: String) {
+        self.0.manufacturer = manufacturer;
+    }
+}
+
+#[derive(Clone)]
+struct ProvisioningService<S> {
+    provisioning_info: Rc<Option<ProvisioningInfo>>,
+    last_connection_attempt: Rc<Option<NetworkInfo>>,
+    reason: Rc<ProvisioningReason>,
+    storage: S,
+    credential_ready: AtomicSync,
+}
+
+impl<S> ProvisioningService<S>
+where
+    S: Storage + Clone,
+    GrpcError: From<S::Error>,
+{
+    async fn process_request_inner(&self, req: Request<Incoming>) -> Result<Bytes, GrpcError> {
+        let (parts, body) = req.into_parts();
+        let mut body = body
+            .collect()
+            .await
+            .map_err(|_| GrpcError::RpcFailedPrecondition)?
+            .to_bytes();
+        match parts.uri.path() {
+            "/viam.provisioning.v1.ProvisioningService/GetSmartMachineStatus" => {
+                self.get_smart_machine_status()
+            }
+            "/viam.provisioning.v1.ProvisioningService/SetSmartMachineCredentials" => {
+                self.set_smart_machine_credentials(body.split_off(5))
+            }
```

Comment:
During what part of the lifetime of a micro-rdk run will these endpoints be available?

Reply:
just during provisioning